README.md
# mandel-rust: Mandelbrot set in Rust

This code shows how to calculate the set in serial and parallel using Rust and various libraries.
More parallel versions (with different libraries) will be added in the future.

Written by Willi Kappler, License: MIT - Version 0.4 (2016.02.28)

![mandelbrot set](mandel.png)


Compile with:

    cargo build --release

Run with the default values:

    cargo run --release

Supported command line options:

        --img_size <IMAGE_SIZE>              size of image in pixel (square, default: 2048, must be a power of two)
        --img1 <IMAGINARY1>                  lower part (default: -1.50)
        --img2 <IMAGINARY2>                  upper part (default: 1.50)
        --write_metadata                     write metadata like run time into the ppm file (default: off)
        --no_ppm                             disable creation of the ppm file, just run the calculation (default: off)
        --bench                              use all available CPUs (default: off), will change in the future
        --max_iter <MAX_ITER>                maximum number of iterations (default: 4096)
        --num_threads <NUMBER_OF_THREADS>    number of threads to use (default: 2)
        --num_of_runs=[NUM_OF_RUNS]          number of repetitive runs (default: 1)
        --re1 <REAL1>                        left real part (default: -2.0)
        --re2 <REAL2>                        right real part (default: 1.0)

The main program runs the calculation 9 times: 1 x single threaded and currently 8 x multi threaded.
It writes the mandelbrot set out as PPM image files. For each method one image file is created.

To check if all the images are equal (and thus that all the computations are correct) you can use this command:

    for i in *.ppm; do md5sum $i; done

Or even better:

    for i in *.ppm; do md5sum $i; done | cut -c1-32 | uniq

(This works only if the flag `--write_metadata` has not been set)

Discussion in the Rust user forum:

- [current version (v0.3)](https://users.rust-lang.org/t/mandel-rust-v0-3-more-crates-more-options/4468)

- [old version (v0.2)](https://users.rust-lang.org/t/new-version-of-mandel-rust-uses-rayon-added-benchmark/4403)



# Used crates:
- [Clap](https://github.com/kbknapp/clap-rs): command line parsing
- [Time](https://github.com/rust-lang-deprecated/time): timing each run
- [Num](https://github.com/rust-num/num): complex numbers
- [Num_cpus](https://github.com/seanmonstar/num_cpus): for the `--bench` flag, determine the total number of cpus
- [Scoped threadpool](https://github.com/Kimundi/scoped-threadpool-rs): use scope and thread pool
- [Simple parallel](https://github.com/huonw/simple_parallel): use scope and thread pool
- [Rayon](https://github.com/nikomatsakis/rayon): using recursive divide-and-conquer call with join, use par_iter_mut
- [Rust scoped pool](https://github.com/reem/rust-scoped-pool): use scope and thread pool
- [Jobsteal](https://github.com/rphmeier/jobsteal): use scope and thread pool, use join (divide-and-conquer). Thanks to Robert Habermeier for the code!
- [Kirk](https://github.com/kinghajj/kirk) + [crossbeam](https://github.com/aturon/crossbeam): use scope and thread pool

# Benchmark:
Measured on a Transtec server with the following specs:
- RAM: 32 GB
- CPU: 2 x Intel Xeon(R) CPU E5-2620 v3 @ 2.40GHz (12 Cores, with hyper threading 24 cores)
- Operating system: 64 bit Ubuntu Server 14.04
- Rust version: rustc 1.6.0 (c30b771ad 2016-01-19)
- Mandel configuration: re1: -2.00, re2: 1.00, img1: -1.50, img2: 1.50, max_iter: 2048, img_size: 1024


![mandelbrot benchmark plot](plot/mandel_bench.png)

(The actual numbers are in the text files under the folder plot/)

I've increased the time to calculate the mandelbot set by making the image size bigger and using a higher iteration.
Thus the differences between the methods become less significant.
Using only one thread the overhead is also less noticeable. simple_parallel has a high variance for thread numbers between 2 and 6.
Here is a closeup for the case where the number of threads are greater than 10:


![mandelbrot benchmark plot closeup](plot/mandel_bench_closeup.png)


As always take these results with a grain of salt, they just show a general direction.
If in doubt just do run some benchmarks with different crates for your specific code (which is always a good thing to do).

# TODO:
- [ ] Check [ArrayFire](https://github.com/arrayfire/arrayfire-rust)
- [ ] Check [Collenchyma](https://github.com/autumnai/collenchyma)
- [ ] Check [Timely Dataflow](https://github.com/frankmcsherry/timely-dataflow)
- [ ] Check [forkjoin](https://github.com/faern/forkjoin)
- [ ] Check [rust-stm](https://github.com/Marthog/rust-stm)
- [ ] Check [mpi](https://github.com/bsteinb/rsmpi)

- [x] Automate benchmark: re-run each test multiple times (user specified command line argument) and take the average
- [x] Automate benchmark: write all results to text files and make a nice plot

- [x] Use a bigger image size and a higher number of iterations for the next release

Any feedback is welcome!


CHANGELOG.md
# Change Log

## 0.4 - 2016-02-28
Refactor utility functions to crate mandel_util and crate mandel_method
New command line option: `--num_of_runs`
Write benchmark results to file
Adapt plot script to read in result file
Add new bench script
Changes default values for max_iter (4096) and img_size (1024)
Add rustc version string at compile time

## 0.3 - 2016-01-30
Added support for rust scoped pool and jobsteal
Added second method for rayon: par_iter
New command line options: `--no-ppm`, `--bench`

## 0.2 - 2016-01-24
Added support for Rayon
Added more information and benchmark in the readme file

## 0.1 - 2015-12-28
First public release on github


src/main.rs
// Mandelbrot set in rust
//
// This code shows how to calculate the set in serial and parallel.
// More parallel versions will be added in the future.
//
// Written by Willi Kappler, grandor@gmx.de
//
// License: MIT
//
//#![feature(plugin)]
//
//#![plugin(clippy)]


// External crates
extern crate time;
extern crate rayon;

// Internal crates
extern crate mandel_util;
extern crate mandel_method;

// External modules
use time::{now};

// Internal modules
use mandel_util::{parse_arguments, do_run, compiler_version};
use mandel_method::*;

fn main() {
    // For example run with:
    // cargo run --release -- --re1=-2.0 --re2=1.0 --img1=-1.5 --img2=1.5
    //           --max_iter=2048 --img_size=1024 --num_threads=2
    //
    // Or just using the default values:
    // cargo run --release -- --num_threads=2
    //
    // Note that the image size must be a power of two

    let mandel_config = parse_arguments();

    let version = env!("CARGO_PKG_VERSION");

    println!("mandel-rust version: {}", version);
    println!("Number of repetitive runs: {}", mandel_config.num_of_runs);
    println!("Rustc version: {}", compiler_version);

    // Get current date and time once and pass it to the individual runs for the image filename.
    let tm = now();
    let tm = tm.strftime("%Y_%m_%d__%H_%M_%S").unwrap();
    let time_now = format!("{}", &tm);

    // vec! macro expects usize
    let mut image: Vec<u32> = vec![0; (mandel_config.img_size * mandel_config.img_size) as usize];

    do_run("serial", &serial, &mandel_config, &mut image, &time_now);

    do_run("scoped_thread_pool", &scoped_thread_pool_, &mandel_config, &mut image, &time_now);

    do_run("simple_parallel", &simple_parallel_, &mandel_config, &mut image, &time_now);

    // Make sure this is only called once
    match rayon::initialize(rayon::Configuration::new().set_num_threads(mandel_config.num_threads as usize)) {
        Ok(_) => {
            do_run("rayon_join", &rayon_join, &mandel_config, &mut image, &time_now);

            do_run("rayon_par_iter", &rayon_par_iter, &mandel_config, &mut image, &time_now);
        },
        Err(e) => println!("Rayon error: set number of threads failed: {}", e)
    }

    do_run("rust_scoped_pool", &rust_scoped_pool, &mandel_config, &mut image, &time_now);

    do_run("job_steal", &job_steal, &mandel_config, &mut image, &time_now);

    do_run("job_steal_join", &job_steal_join, &mandel_config, &mut image, &time_now);

    do_run("kirk_crossbeam", &kirk_crossbeam, &mandel_config, &mut image, &time_now);
}


mandel_method/src/lib.rs
// External crates
extern crate num;
extern crate scoped_threadpool;
extern crate simple_parallel;
extern crate rayon;
extern crate scoped_pool;
extern crate jobsteal;
extern crate kirk;
extern crate crossbeam;

// Internal crates
extern crate mandel_util;

// External modules
use num::complex::Complex64;
use rayon::par_iter::*;
use kirk::crew::deque::Options;

// Internal modules
use mandel_util::{mandel_iter, MandelConfig};

// The serial version of the mandelbrot set calculation.
pub fn serial(mandel_config: &MandelConfig, image: &mut [u32]) {
    for y in 0..mandel_config.img_size {
        for x in 0..mandel_config.img_size {
            image[((y * mandel_config.img_size) + x) as usize] =
                mandel_iter(mandel_config.max_iter,
                    Complex64{re: mandel_config.re1 + ((x as f64) * mandel_config.x_step),
                              im: mandel_config.img1 + ((y as f64) * mandel_config.y_step)}
                );
        }
    }
}

// The parallel version of the mandelbrot set calculation, uses scoped_threadpool.
pub fn scoped_thread_pool_(mandel_config: &MandelConfig, image: &mut [u32]) {
    let mut pool = scoped_threadpool::Pool::new(mandel_config.num_threads);

    pool.scoped(|scope| {
        for (y, slice) in image.chunks_mut(mandel_config.img_size as usize).enumerate() {
            scope.execute(move || {
                for x in 0..mandel_config.img_size {
                    slice[x as usize] =
                    mandel_iter(mandel_config.max_iter,
                        Complex64{re: mandel_config.re1 + ((x as f64) * mandel_config.x_step),
                                  im: mandel_config.img1 + ((y as f64) * mandel_config.y_step)}
                    );
                }
            });
        }
    });
}

// The parallel version of the mandelbrot set calculation, uses simple_parallel.
pub fn simple_parallel_(mandel_config: &MandelConfig, image: &mut [u32]) {
    let mut pool = simple_parallel::Pool::new(mandel_config.num_threads as usize);

    pool.for_(image.chunks_mut(mandel_config.img_size as usize).enumerate(), |(y, slice)| {
        for x in 0..mandel_config.img_size {
            slice[x as usize] =
            mandel_iter(mandel_config.max_iter,
                Complex64{re: mandel_config.re1 + ((x as f64) * mandel_config.x_step),
                          im: mandel_config.img1 + ((y as f64) * mandel_config.y_step)}
            );
        }
    });
}

// The parallel version of the mandelbrot set calculation, uses rayon join.
pub fn rayon_join(mandel_config: &MandelConfig, image: &mut [u32]) {
    rayon_helper(mandel_config, image, 0);
}

// Rayon helper function for recursive divide-and-conquer call
fn rayon_helper(mandel_config: &MandelConfig, slice: &mut [u32], y: u32) {
    if slice.len() == (mandel_config.img_size as usize) { // just process one scanline of the mandelbrot image
        for x in 0..mandel_config.img_size {
            slice[x as usize] =
            mandel_iter(mandel_config.max_iter,
                Complex64{re: mandel_config.re1 + ((x as f64) * mandel_config.x_step),
                          im: mandel_config.img1 + ((y as f64) * mandel_config.y_step)}
            );
        }
    } else {
        let mid = slice.len() / 2;
        let (top, bottom) = slice.split_at_mut(mid);
        rayon::join(
            || rayon_helper(mandel_config, top, y),
            || rayon_helper(mandel_config, bottom, y + ((mid as u32) / mandel_config.img_size))
        );
    }
}

// The parallel version of the mandelbrot set calculation, uses rayon par_iter_mut.
pub fn rayon_par_iter(mandel_config: &MandelConfig, image: &mut [u32]) {

    image.par_iter_mut().enumerate().for_each(
        |(n, pixel)| {
            let y = (n as u32) / mandel_config.img_size;
            let x = (n as u32) - (y * mandel_config.img_size);
            *pixel = mandel_iter(mandel_config.max_iter,
                        Complex64{re: mandel_config.re1 + ((x as f64) * mandel_config.x_step),
                                  im: mandel_config.img1 + ((y as f64) * mandel_config.y_step)}
            );
        });
}

// The parallel version of the mandelbrot set calculation, uses rust scoped pool.
pub fn rust_scoped_pool(mandel_config: &MandelConfig, image: &mut [u32]) {
    let pool = scoped_pool::Pool::new(mandel_config.num_threads as usize);

    pool.scoped(|scope| {
        for (y, slice) in image.chunks_mut(mandel_config.img_size as usize).enumerate() {
            scope.execute(move || {
                for x in 0..mandel_config.img_size {
                    slice[x as usize] =
                    mandel_iter(mandel_config.max_iter,
                        Complex64{re: mandel_config.re1 + ((x as f64) * mandel_config.x_step),
                                  im: mandel_config.img1 + ((y as f64) * mandel_config.y_step)}
                    );
                }
            });
        }
    });
}

// The parallel version of the mandelbrot set calculation, uses jobsteal.
pub fn job_steal(mandel_config: &MandelConfig, image: &mut [u32]) {
    let mut pool = jobsteal::make_pool((mandel_config.num_threads - 1) as usize).unwrap();

    pool.scope(|scope| {
        for (y, slice) in image.chunks_mut(mandel_config.img_size as usize).enumerate() {
            scope.submit(move || {
                for x in 0..mandel_config.img_size {
                    slice[x as usize] =
                    mandel_iter(mandel_config.max_iter,
                        Complex64{re: mandel_config.re1 + ((x as f64) * mandel_config.x_step),
                                  im: mandel_config.img1 + ((y as f64) * mandel_config.y_step)}
                    );
                }
            });
        }
    });
}

// The parallel version of the mandelbrot set calculation, uses jobsteal with divide-and-conquer strategy.
pub fn job_steal_join(mandel_config: &MandelConfig, image: &mut [u32]) {
    // Jobsteal uses n + 1 threads (1 main thread + n sub-threads)
    // It is OK to create a Jobsteal pool with zero threads.
    // See https://github.com/willi-kappler/mandel-rust/issues/1
    let mut pool = jobsteal::make_pool((mandel_config.num_threads - 1) as usize).unwrap();

    pool.scope(|scope| {
        job_steal_helper(mandel_config, scope, image, 0);
    })
}

// jobsteal helper for divide and conquer version.
fn job_steal_helper<'a, 'b>(mandel_config: &MandelConfig, spawner: &jobsteal::Spawner<'a, 'b>,
                            slice: &mut [u32], y: u32) {
    if slice.len() == (mandel_config.img_size as usize) { // just process one scanline of the mandelbrot image
        for x in 0..mandel_config.img_size {
            slice[x as usize] =
            mandel_iter(mandel_config.max_iter,
                Complex64{re: mandel_config.re1 + ((x as f64) * mandel_config.x_step),
                          im: mandel_config.img1 + ((y as f64) * mandel_config.y_step)}
            );
        }
    } else {
        let mid = slice.len() / 2;
        let (top, bottom) = slice.split_at_mut(mid);
        spawner.join(
            |inner| job_steal_helper(mandel_config, inner, top, y),
            |inner| job_steal_helper(mandel_config, inner, bottom, y + ((mid as u32) / mandel_config.img_size))
        );
    }
}

// The parallel version of the mandelbrot set calculation, uses kirk and crossbeam.
pub fn kirk_crossbeam(mandel_config: &MandelConfig, image: &mut [u32]) {
    crossbeam::scope(|scope| {
        let mut pool = kirk::Pool::<kirk::Deque<kirk::Task>>::scoped(scope,
            Options{ num_workers: mandel_config.num_threads as usize, .. Options::default()});
        for (y, slice) in image.chunks_mut(mandel_config.img_size as usize).enumerate() {
            pool.push(move || {
                for x in 0..mandel_config.img_size {
                    slice[x as usize] =
                    mandel_iter(mandel_config.max_iter,
                        Complex64{re: mandel_config.re1 + ((x as f64) * mandel_config.x_step),
                                  im: mandel_config.img1 + ((y as f64) * mandel_config.y_step)}
                    );
                }
            })
        }
    });
}


src/mandel_util.rs

// External crates
#[macro_use]
extern crate clap;
extern crate num_cpus;
extern crate num;
extern crate time;

// External modules
use clap::App;
use num::complex::Complex64;
use time::{precise_time_ns};

// Rust modules
use std::fs::File;
use std::io::prelude::Write;
use std::io::Result;
use std::io::BufWriter;
use std::fs::OpenOptions;

// Configuration file, reflects command line options
#[derive(Copy, Clone)]
pub struct MandelConfig {
    pub re1: f64,
    pub re2: f64,
    pub img1: f64,
    pub img2: f64,
    pub x_step: f64,
    pub y_step: f64,
    pub max_iter: u32,
    pub img_size: u32,
    pub write_metadata: bool,
    pub no_ppm: bool,
    pub num_threads: u32,
    pub num_of_runs: u32
}

include!(concat!(env!("OUT_DIR"), "/compiler_version.rs"));

// Parse command line options via clap and returns the responding configuration
pub fn parse_arguments() -> MandelConfig {
    let matches = App::new("mandel_rust")
        .version("0.3")
        .author("Willi Kappler <grandor@gmx.de>")
        .about("Simple mandelbrot written in pure rust")
        .args_from_usage(
            "--re1=[REAL1] 'left real part (default: -2.0)'
             --re2=[REAL2] 'right real part (default: 1.0)'
             --img1=[IMAGINARY1] 'lower part (default: -1.50)'
             --img2=[IMAGINARY2] 'upper part (default: 1.50)'
             --write_metadata 'write metadata like run time into the ppm file (default: off)'
             --no_ppm 'disable creation of the ppm file, just run the calculation (default: off)'
             --bench 'use all available CPUs (default: off), will change in the future'
             --max_iter=[MAX_ITER] 'maximum number of iterations (default: 4096)'
             --img_size=[IMAGE_SIZE] 'size of image in pixel (square, default: 2048, must be a power of two)'
             --num_of_runs=[NUM_OF_RUNS] 'number of repetitive runs (default: 1)'
             --num_threads=[NUMBER_OF_THREADS] 'number of threads to use (default: 2)'")
        .get_matches();

    let re1 = value_t!(matches.value_of("REAL1"), f64).unwrap_or(-2.0);
    let re2 = value_t!(matches.value_of("REAL2"), f64).unwrap_or(1.0);
    let img1 = value_t!(matches.value_of("IMAGINARY1"), f64).unwrap_or(-1.5);
    let img2 = value_t!(matches.value_of("IMAGINARY2"), f64).unwrap_or(1.5);
    let metadata = matches.is_present("write_metadata");
    let bench = matches.is_present("bench");
    let no_ppm = matches.is_present("no_ppm");
    let max_iter = value_t!(matches.value_of("MAX_ITER"), u32).unwrap_or(4096);
    let img_size = value_t!(matches.value_of("IMAGE_SIZE"), u32).unwrap_or(2048);
    let num_of_runs = value_t!(matches.value_of("NUM_OF_RUNS"), u32).unwrap_or(1);
    let num_threads = if bench { num_cpus::get() as u32 } else {
        value_t!(matches.value_of("NUMBER_OF_THREADS"), u32).unwrap_or(2) };

    assert!(re1 < re2);
    assert!(img1 < img2);
    assert!(max_iter > 0);
    assert!(img_size > 0);
    assert!(num_threads > 0);

    println!("Configuration: re1: {:.2}, re2: {:.2}, img1: {:.2}, img2: {:.2}, max_iter: {}, img_size: {}, num_threads: {}",
        re1, re2, img1, img2, max_iter, img_size, num_threads);

    let x_step = (re2 - re1) / (img_size as f64);
    let y_step = (img2 - img1) / (img_size as f64);

    MandelConfig{
        re1: re1,
        re2: re2,
        img1: img1,
        img2: img2,
        x_step: x_step,
        y_step: y_step,
        max_iter: max_iter,
        img_size: img_size,
        write_metadata: metadata,
        no_ppm: no_ppm,
        num_threads: num_threads,
        num_of_runs: num_of_runs
    }
}

// The inner iteration loop of the mandelbrot calculation
// See https://en.wikipedia.org/wiki/Mandelbrot_set
pub fn mandel_iter(max_iter: u32, c: Complex64) -> u32 {
    let mut z: Complex64 = c;

    let mut iter = 0;

    while (z.norm_sqr() <= 4.0) && (iter < max_iter) {
        z = c + (z * z);
        iter = iter + 1;
    }

    iter
}

// Write calculated mandelbrot set as PPM image.
// Add run time information as comment.
fn write_image(file_name: &str, mandel_config: &MandelConfig, time_in_ms: f64, image: &[u32]) -> Result<()> {
    let mut buffer = BufWriter::new(try!(File::create(file_name)));

    try!(buffer.write(b"P3\n"));
    try!(write!(buffer, "# mandelbrot, max_iter: {}\n", mandel_config.max_iter));
    if mandel_config.write_metadata {
        // TODO: add more meta data: date and time, method, ...
        try!(write!(buffer, "# computation time: {} ms\n", time_in_ms));
    }
    try!(write!(buffer, "{0} {0}\n", mandel_config.img_size));
    try!(buffer.write(b"255\n"));

    let mut img_value: u32;

    for y in 0..mandel_config.img_size {
        for x in 0..mandel_config.img_size {

            img_value = image[((y * mandel_config.img_size) + x) as usize];
            if img_value == mandel_config.max_iter {
                try!(buffer.write(b"0 0 0 "));
            } else {
                try!(write!(buffer, "255 {} 0 ", (img_value % 16) * 16));
            }

        }
        try!(buffer.write(b"\n"));
    }

    Ok(())
}

fn write_benchmark_result(method: &str, num_threads: u32,
    time_in_ms: f64, min_time: f64, max_time: f64) -> Result<()> {
    let mut buffer = BufWriter::new(try!(
        OpenOptions::new()
            .write(true)
            .append(true)
            .create(true)
            .open(format!("plot/{}.txt", method))));

    try!(write!(buffer, "{} {} {} {}\n", num_threads, time_in_ms, min_time, max_time));

    Ok(())
}

// Prepares and runs one version of the mandelbrot set calculation.
pub fn do_run(method: &str, mandel_func: &Fn(&MandelConfig, &mut [u32]) -> (),
    mandel_config: &MandelConfig, image: &mut [u32], time_now: &str) {

    let mut repetitive_times = Vec::new();
    let mut min_time = std::f64::MAX;
    let mut max_time = 0.0;

    for _ in 0..mandel_config.num_of_runs {
        let start_time = precise_time_ns();

        mandel_func(mandel_config, image);

        let end_time = precise_time_ns();
        let total_time_in_ms = ((end_time - start_time) as f64) / (1000.0 * 1000.0);

        if total_time_in_ms > max_time {
            max_time = total_time_in_ms;
        } else if total_time_in_ms < min_time {
            min_time = total_time_in_ms;
        }

        repetitive_times.push(total_time_in_ms);
    }

    let mean_time = repetitive_times.iter().fold(0.0, |sum, t| sum + t) /
        (mandel_config.num_of_runs as f64);

    println!("Time taken for this run ({}): {:.5} ms", method, mean_time);

    write_benchmark_result(&method, mandel_config.num_threads, mean_time,
        min_time, max_time).expect("I/O error while writing benchmark results");

    if !mandel_config.no_ppm {
        let file_name = format!("{}_{}.ppm", method, &time_now);

        write_image(&file_name, &mandel_config, mean_time, &image).expect(
            &format!("I/O error while writing image: '{}'", file_name));
    }
}
